hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -283,8 +287,11 @@ protected void commit(HoodieTable table, String commitActionType, String instant
     }
     // update Metadata table
     writeTableMetadata(table, instantTime, metadata, writeStatuses);
-    activeTimeline.saveAsComplete(new HoodieInstant(true, commitActionType, instantTime),
-        Option.of(metadata.toJsonString().getBytes(StandardCharsets.UTF_8)));
+    activeTimeline.saveAsComplete(
+        new HoodieInstant(true, commitActionType, instantTime),
+        metadata instanceof HoodieReplaceCommitMetadata
+            ? serializeReplaceCommitMetadata(convertReplaceCommitMetadata((HoodieReplaceCommitMetadata) metadata))
+            : serializeCommitMetadata(convertCommitMetadata(metadata)));
   }
```

Comment:
I see the nested invocation of `serializeReplaceCommitMetadata(convertReplaceCommitMetadata(` everywhere, if that is alway a combination, I suggest we move the conversion `convertReplaceCommitMetadata` inside the method `serializeReplaceCommitMetadata`.

Also we can remove `serializeReplaceCommitMetadata` and fuse its logic into `serializeCommitMetadata`.

Reply:
Done.